### ISSUE TITLE
Better to use environment concept

### ISSUE THREAD

**Issue by seajean** [alibaba/weex#1649]
Currently there is only instance concept,  every instance is independent, and it's difficult to share data between instances.

## Recommendation
import environment concept,  before creating instance, create environment first, 
after create an environment,  js framework returns  an environment ID, and then we can create an instance inside the specified environment. 

- developer can specify JS framework version which include  build-in-libs version.
- Adding new function like  `createEnv`. 
- Adding param envID to createInstance function.
 

## why we need these feature

- Our js framework and build-in-libs are being developed all the time, some instance code may not compatible with the js framework we are using now,  thus developers must modify their code to adapt to our js framework. with environment , developer could specify  which version of js framework and build-in-libs their are using, and before creating instance, we can create an environment for that instance.

-  Considering  another situation, several page was developed by  the same developer, and it is very  common to share data between pages.  developer could share data inside the environment with no harm to the others. Developer can declare  an environment variable and use it inside that environment.

**Comment by seajean** [comment]
Another problem problem we are facing now is that too much thing was done when creating an instance,  if we prepare the environment first and the creating time will be much shorter.

**Comment by Jinjiang** [comment]
If the framework code is not stable enough, I have two advices:

1. make the update as backward compatible as possible.
2. do not put the unstable code into the framework.

And I'm not quite sure about what `createEnv` will do. It's better to have more specific example if you don't mind.

The second question is about inter-instance communications. We had another issue #1536 and implementation #1644 about this in discussion.

Thanks.

**Comment by seajean** [comment]
Maybe the framework can be backward compatible, but not the build-in-lib,  With environment, we can split our code into 3 parts, the js-framework, the build-in-lib, then the  instance code, the build-in-lib may changed over time and have many versions.Developer can specify which version of build-in-lib they want to use.

createENV is some code like below

```javascript
var env = {};

function creataENV(id){
  var a = 1;
  var b = 2;
  var def = function() {
    
  }
  var sdk1 = function(){
     return '<div> sdk ' + id + '</div>';
  }
 
  env[id] =  {
     def: def,
     a: a,
     b: b,
     sdk1: sdk1
  }
  return id;
}
let code = 'var sdk3 = sdk1; document.write(sdk3())';

function createInstance(id, d) {
   let fn = new Function('def','a', 'sdk1', d);
   fn(env[id].def, env[id].a, env[id].sdk1);
}


var env1 = creataENV('env1');
createInstance(env1, code);

var env1 = creataENV('env2');
createInstance(env1, code);
```

**Comment by Hanks10100** [comment]
Thanks for the suggestion. Currently weex has a concept of `runtime`, which is similar with your `environment`. The `runtime` is created when the weex sdk is initialized. All the instances are created in the `runtime` context, each instances has an `id, and the singleton `runtime` is responsible for managing them.

If you want to achieve preloading the built-in library, may be #1514 is related.

I personally think that each environment has only one particular version of js framework is more reasonable, otherwise the complexity will increase significantly. However, we support to use multiple "cores", such as *Vue 2.0* and our own *js framework*, but the version is fixed.

**Comment by seajean** [comment]
Currently we have manage to achieve  preloading build-in-lib,  and  this suggestion was  thrown out when we found that  runtime still  is not good enough.

**Comment by Hanks10100** [comment]
We are now thinking about how to solve "data management" problems, which may be not simply adding `environment` or communication mechanisms, it should be more progressive.

Weex is a cross platform framework, its development experience is not limited to the front-end. The data management should have an elegant solution, such as Flux, RxJS, etc (they may not appropriate). Feel free to provide your perspective.

**Comment by seajean** [comment]
It's not just data management, but build-in-lib version control. jsfm and bundle two layer architecture is not fit to do build-in-lib version control. so we may need the environment, the third layer.

**Comment by Hanks10100** [comment]
> I personally think that each environment has only one particular version of js framework is more reasonable.

We certainly have a reason to upgrade the version, maybe it's a bugfix. So far we haven't released 1.0 or LTS version, we can't guarantee that each version of the js framework are feature-completed.

**Comment by Jinjiang** [comment]
I have some more point of views.

* The requirement I see is:
  1. be able to prepare different (even the same but different version) built-in libs for each instance and
  2. avoid download them each time
* As you mentioned in #1514 those libs may disturb each other or not work well for multi-instances in the same time (or lead memory leaks etc). So it's better to run when in a special time such as `createInstance`.
* We actually should do is able to run/bootstrap different code in each instance by something (for example envID as your proposal).
* Additionally there is another potential user requirement: be able to detect whether a lib is supported now, even with a certain lib version.

I think we could make a global static extension lib/class collections named `__weex_libs__`. To make sure each of them totally static (be able to work separately) and frozen. And the app builder can put them in a whole file and concat with `main.js` (JS framework) together.
If you want to have different versions of just one lib, or if you want to get a sub collection of all extensions, that's all easy and safe to do. This job can be done in JS bundle level with a little code first.

**Comment by seajean** [comment]
I am afraid  doing too much  things at the same time  that  the app may hang up for a while, such as   initializing the big size built-in-libs and launching the app or  creating the instance  at the same time.  We initialize the js-framework right after the app launched,  so put the built-in-libs' code into the js-framework is not a good idea as the built-in-libs' size is a little big and have many versions.  We are thinking about  to load and execute built-in-libs  only after  user clicked  the page link.
The question is  not only  where we put our built-in-libs' code but also when we execute it,built-in-libs is  just too big too push them in the js-framework.js or the bundle.js.

**Comment by Jinjiang** [comment]
I think the size of JS framework is not a big problem in native if the libs aren't executed (just defined or prepared a little). If we want to make the libs safe, it must be more chosen and executed when we creating instance.
So my suggestion is concat with JS framework (with some preparation) and execute by deps in JS bundle when createInstance.

**Comment by seajean** [comment]
My solution for the built-in-libs' small  updates  is put them in cdn and check for updates each time when  the app launch, is there an alternative way to solve this?   something like CodePush?

**Comment by Jinjiang** [comment]
After several discussions we recommend JS Service #1938 
Maybe we can move on there
Thanks.
